NetKet/Graph/pygraph.hpp

```diff
@@ -99,62 +138,55 @@ auto WithEdges(py::iterable xs, Function&& callback)
-        throw InvalidInputError{"Edge list contains duplicates."};
-      }
-    }
+    auto colors = detail::Iterable2ColorMap(xs);
+    std::vector<AbstractGraph::Edge> edges;
+    edges.reserve(colors.size());
+    std::transform(
+        begin(colors), end(colors), std::back_inserter(edges),
+        [](std::pair<AbstractGraph::Edge, int> const& x) { return x.first; });
     return std::forward<Function>(callback)(std::move(edges),
-                                            std::move(colours));
+                                            std::move(colors));
+  } else {
+    return std::forward<Function>(callback)(detail::Iterable2Edges(xs));
   }
-  // No colours
-  for (; first != last; ++first) {
-    auto const x = first->template cast<std::tuple<int, int>>();
-    edges.push_back(make_edge(std::get<0>(x), std::get<1>(x)));
-  }
-  // NOTE(twesterhout): Yes, I know that this screws up the algorithmic
-  // complexity, but it's fast enough to be unnoticeable for all practical
-  // purposes.
-  std::sort(begin(edges), end(edges));
-  if (std::unique(begin(edges), end(edges)) != end(edges)) {
-    throw InvalidInputError{"Edge list contains duplicates."};
-  }
-  return std::forward<Function>(callback)(std::move(edges));
 }
+} // namespace
 
 void AddGraphModule(py::module& m) {
   auto subm = m.def_submodule("graph");
 
   py::class_<AbstractGraph>(subm, "Graph") ADDGRAPHMETHODS(AbstractGraph);
 
   py::class_<Hypercube, AbstractGraph>(subm, "Hypercube")
-
-      .def(py::init<int, int, bool, std::vector<std::vector<int>>>(),
-           py::arg("L"), py::arg("ndim"), py::arg("pbc") = true,
-           py::arg("edgecolors") = std::vector<std::vector<int>>())
-      .def("Nsites", &Hypercube::Nsites)
-      .def("AdjacencyList", &Hypercube::AdjacencyList)
-      .def("SymmetryTable", &Hypercube::SymmetryTable)
-      .def("EdgeColors", &Hypercube::EdgeColors)
-      .def("IsBipartite", &Hypercube::IsBipartite)
-      .def("IsConnected", &Hypercube::IsConnected)
-      .def("Distances", &Hypercube::Distances)
-      .def("AllDistances", &Hypercube::AllDistances) ADDGRAPHMETHODS(Hypercube);
+      .def(py::init<int, int, bool>(), py::arg("length"), py::arg("ndim") = 1,
+           py::arg("pbc") = true)
+      .def(py::init([](int const length, py::iterable xs) {
+             return make_unique<Hypercube>(length,
+                                           detail::Iterable2ColorMap(xs));
+           }),
+           py::arg("length"), py::arg("colors"))
+      .def_property_readonly("number_sites", &Hypercube::Nsites)
+      .def("__len__", &Hypercube::Nsites)
+      .def("__iter__",
+           [](Hypercube const& x) {
+             using std::begin;
+             using std::end;
+             return py::make_iterator(begin(x.Edges()), end(x.Edges()));
+           },
+           py::keep_alive<0, 1>())
+      // NOTE(twesterhout): This is redundant, but I'm not yet sure treating
+      // graph as an iterable over edges is a good idea...
```

Comment:
Iterating over the graph directly could sensibly refer to iterating over vertices or edges, so IMHO it's best to force calling code to be explicit and call `.edges`.

Reply:
Great, thanks!